src/restrictededitingmodeediting.js

```diff
@@ -0,0 +1,275 @@
+		 *
+		 * @type {Set<string>}
+		 * @private
+		 */
+		this._alwaysEnabled = new Set( [ 'undo', 'redo', 'goToPreviousRestrictedEditingRegion', 'goToNextRestrictedEditingRegion' ] );
+
+		/**
+		 * Commands allowed in non-restricted areas.
+		 *
+		 * Commands always enabled combines typing feature commands: `'typing'`, `'delete'` and `'forwardDelete'` with commands defined
+		 * in the feature configuration.
+		 *
+		 * @type {Set<string>}
+		 * @private
+		 */
+		this._allowedInException = new Set( [ 'input', 'delete', 'forwardDelete' ] );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		const editor = this.editor;
+
+		const allowedCommands = editor.config.get( 'restrictedEditing.allowedCommands' );
+
+		allowedCommands.forEach( commandName => this._allowedInException.add( commandName ) );
+
+		this._setupConversion();
+		this._setupCommandsToggling();
+
+		// Commands & keystrokes that allow navigation in the content.
+		editor.commands.add( 'goToPreviousRestrictedEditingRegion', new RestrictedEditingNavigationCommand( editor, 'backward' ) );
+		editor.commands.add( 'goToNextRestrictedEditingRegion', new RestrictedEditingNavigationCommand( editor, 'forward' ) );
+		editor.keystrokes.set( 'Tab', getCommandExecuter( editor, 'goToNextRestrictedEditingRegion' ) );
+		editor.keystrokes.set( 'Shift+Tab', getCommandExecuter( editor, 'goToPreviousRestrictedEditingRegion' ) );
+
+		// Block clipboard completely in restricted mode.
+		this.listenTo( this.editor.editing.view.document, 'clipboardInput', evt => {
+			evt.stop();
+		}, { priority: 'highest' } );
+	}
+
+	/**
+	 * Setups restricted mode editing conversion:
+	 *
+	 * * ucpast & downcast converters
+	 * * marker highlighting in the edting area
+	 * * marker post-fixers
+	 *
+	 * @private
+	 */
+	_setupConversion() {
+		const editor = this.editor;
+		const model = editor.model;
+		const doc = model.document;
+
+		// The restricted editing does not attach additional data to the zones so there's no need for smarter markers management.
+		// Also, the markers will only be created when  when loading the data.
+		let markerNumber = 0;
+
+		editor.conversion.for( 'upcast' ).add( upcastHighlightToMarker( {
+			view: {
+				name: 'span',
+				classes: 'ck-restricted-editing-exception'
+			},
+			model: () => {
+				markerNumber++; // Starting from restricted-editing-exception:1 marker.
+
+				return `restricted-editing-exception:${ markerNumber }`;
+			}
+		} ) );
+
+		// Currently the marker helpers are tied to other use-cases and do not render collapsed marker as highlight.
+		// That's why there are 2 downcast converters for them:
+		// 1. The default marker-to-highlight will wrap selected text with `<span>`.
+		editor.conversion.for( 'downcast' ).markerToHighlight( {
+			model: 'restricted-editing-exception',
+			// Use callback to return new object every time new marker instance is created - otherwise it will be seen as the same marker.
+			view: () => ( {
```

Comment:
TBH, I prefer:

```js
view() {
     return { 
         foo: 1,
         bar: 2
     };
}
```

Not a big deal but I think that great majority of currently existing code follows that. There are only 7 occurrences of `() => ( {`.

Reply:
But we can ask more people for their opinions.

Note: below there's another fragment that I'd rather change:

```js
editor.conversion.for( 'dataDowncast' ).markerToElement( {
	model: 'restricted-editing-exception',
	view: ( markerData, viewWriter ) => viewWriter.createEmptyElement( 'span', {
		class: 'ck-restricted-editing-exception'
	} )
} );
```